library/L1_Peripheral/lpc40xx/eeprom.hpp

```diff
@@ -0,0 +1,163 @@
+
+    // Writing a 32-bit number to an address that isn't a multiple of 4 will
+    // cause a hard fault, thus this safety is put into place to avoid that.
+    full_address &= 0b1111'1111'1100;
+
+    // Because the EEPROM uses 32-bit communication, write_data will be casted
+    // into a uint32_t *
+    const uint32_t * write_data = reinterpret_cast<const uint32_t *>(wdata);
+
+    // The first 6 bits in the address (MSB) dictate which page is being written
+    // to in the EEPROM, and the last 6 bits (LSB) dictate offset in the page
+    uint8_t page_count  = (full_address & 0b1111'1100'0000) >> 6;
+    uint8_t page_offset = full_address & 0b0000'0011'1111;
+
+    // Page offset is incremented by 4 every time because we're writing 32 bits
+    uint8_t offset_interval = 4;
+
+    uint16_t address;
+
+    while (count > 0)
+    {
+      address = static_cast<uint16_t>((page_count << 6) + page_offset);
+
+      eeprom_register->ADDR  = address;
+      eeprom_register->CMD   = kWrite32Bits;
+      eeprom_register->WDATA = write_data[0];
+
+      write_data++;
+      page_offset = static_cast<uint8_t>(page_offset + offset_interval);
+      count -= 4;
+
+      // If the 64 byte page buffer fills up, then it must be programmed to the
+      // EEPROM before continuing.
+      if (page_offset > kMax6Bits)
+      {
+        Program(address);
+        page_count++;
+        page_offset = 0;
+      }
+    }
+
+    // Program final information so that it's stored in the EEPROM
+    Program(address);
+  }
+
+  // The EEPROM is accessed through a 64 byte page buffer, and after it fills
+  // up, it must be programmed to the actual EEPROM. This function handles that.
+  void Program(uint32_t address) const
+  {
+    eeprom_register->ADDR  = address;
+    eeprom_register->CMD  = kEraseProgram;
+
+    // Poll status register bit to see when programming is finished
+    while ((!(eeprom_register->INT_STATUS >> kProgramStatusBit)) & 1)
+    {
+    };
+
+    // Clear program interrupt
+    eeprom_register->INT_CLR_STATUS = (1 << kProgramStatusBit);
+  }
+
+  // This function will return however much 32-bit data from the EEPROM
+  // starting at address and continuing on for how large count is. The EEPROM
+  // can only be interfaced in 32-bit mode, which means that the address used
+  // must be a multiple of 4. The address entered will be cut down to a
+  // multiple of 4.
+  void Read(uint8_t * rdata, uint32_t address, size_t count) const override
+  {
+    // Writing a 32-bit number to an address that isn't a multiple of 4 will
+    // cause a hard fault, thus this safety is put into place to avoid that.
+    address &= 0b1111'1111'1100;
+
+    // Because the EEPROM uses 32-bit communication, read_data will be casted
+    // into a uint32_t *
+    uint32_t * read_data = reinterpret_cast<uint32_t *> (rdata);
+
+    uint16_t index = 0;
+
+    while ((index * 4) < count)
+    {
+      eeprom_register->ADDR = address;

```

Comment:
The address can be computed using the index here, which allows this to be a for loop.
```suggestion
      eeprom_register->ADDR = address + (index * 4);
```